Lib/test/test_warnings/__init__.py

```diff
@@ -887,37 +888,37 @@ def test_issue31285(self):
         # warn_explicit() should neither raise a SystemError nor cause an
         # assertion failure, in case the return value of get_source() has a
         # bad splitlines() method.
-        def get_bad_loader(splitlines_ret_val):
-            class BadLoader:
-                def get_source(self, fullname):
-                    class BadSource(str):
-                        def splitlines(self):
-                            return splitlines_ret_val
-                    return BadSource('spam')
-            return BadLoader()
-
+        class BadLoader:
+            def get_source(self, fullname):
+                class BadSource(str):
+                    def splitlines(self):
+                        return splitlines_ret_val
+                return BadSource('spam')
+
+        loader = BadLoader()
+        spec = importlib.machinery.ModuleSpec('foobar', loader)
+        module_globals = {'__loader__': loader,
+                          '__spec__': spec,
+                          '__name__': 'foobar'}
```

Comment:
I quite like the `get_bad_loader()` function on `main`, because it makes two things clear to a reader of the test:

1. The only thing different between the two instantiations of `BadLoader` in this test is the return value of `BadSource.splitlines()`
2. It's obvious to the reader where the `splitlines_ret_val` value is set; you don't have to scroll down the test to figure it out

What about having something like this as a local function in the test?

```py
def get_module_globals(*, splitlines_ret_val):
    class BadSource(str):
        def splitilines(self):
            return splitlines_ret_val

    class BadLoader:
        def getsource(self, fullname):
            return BadSource('spam')

    loader = BadLoader()
    spec = importlib.machinery.ModuleSpec('foobar', loader)
    return {'__loader__': loader,
            '__spec__': spec,
            '__name__': 'foobar'}
```

And then the two `wmod.warn_explicit()` calls lower down could be rewritten as (respectively)

```py
wmod.warn_explicit(
    'foo', UserWarning, 'bar', 1,
    module_globals=get_module_globals(splitlines_ret_val=42)
)

wmod.warn_explicit(
    'foo', UserWarning, 'bar', 1,
    module_globals=get_module_globals(splitlines_ret_val=[42])
)
```

Reply:
I get rid of it because it allows to share `module_globals` between tests. Your suggestion is other solution of this problem.

There are other minor disadvantages of `get_bad_loader()` -- it adds a line and additional indentation. But if you like this approach, I'll use it.